File: ene-jun-2023/Jorge_Saucedo/Practica7.py
def quicksort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quicksort(left) + middle + quicksort(right)


def binary_search(arr, target):
    low = 0
    high = len(arr) - 1
    while low <= high:
        mid = (low + high) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
    return -1


# Lectura de la entrada
n, q = map(int, input("Ingresa el numero de medicamentos y el numero de peticiones separadas de un espacio\n").split())
medicinas = input("\nIngresa el nombre de medicamentos separados por un espacio\n").split()

# Ordenar las medicinas
medicinas = quicksort(medicinas)

print("\n**si tu medicamento no existe se regresara un -1**\n")
# Realizar las búsquedas
for _ in range(q):
    query = input(f"Ingresa el {_+1} medicamento a buscar\n").strip()
    position = binary_search(medicinas, query)
    print(f"\n*El medicamento {query} se encuentra en la posicion {position}*\n")

File: ene-jun-2023/Alfonso Hernandez/Practica_7.java
package Algoritmos_Ordenamiento_Busqueda;

import java.util.*;

public class Practica_7 {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(); // número de medicinas
        int q = sc.nextInt(); // número de peticiones

        // Leer los nombres de las medicinas
        String[] medicinas = new String[n];
        for (int i = 0; i < n; i++) {
            medicinas[i] = sc.next();
        }

        // Ordenar el arreglo de medicinas usando quicksort
        quicksort(medicinas, 0, n - 1);

        // Para cada petición, buscar el nombre de la medicina usando binary search
        for (int i = 0; i < q; i++) {
            String nombreMedicina = sc.next();
            int posicion = binarySearch(medicinas, nombreMedicina);
            System.out.println(posicion);
        }
    }

    // Algoritmo quicksort para ordenar un arreglo de strings
    public static void quicksort(String[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);

            quicksort(arr, low, pi - 1);
            quicksort(arr, pi + 1, high);
        }
    }

    // Algoritmo partition para quicksort
    public static int partition(String[] arr, int low, int high) {
        String pivot = arr[high];
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (arr[j].compareTo(pivot) < 0) {
                i++;

                String temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        String temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    // Algoritmo binary search para buscar un string en un arreglo ordenado
    public static int binarySearch(String[] arr, String x) {
        int left = 0;
        int right = arr.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (arr[mid].equals(x)) {
                return mid;
            } else if (arr[mid].compareTo(x) < 0) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }
}
